File: leetcode/2682/find_losers_circular_game.py
# 51ms | 16.52mb
# 49.81% | 78.33%
class Solution:
    def circularGameLosers(self, n: int, k: int) -> list[int]:
        winners = [False for x in range(n)]
        current_friend = 0
        current_k = 1

        playing = True
        while playing:
            if winners[current_friend]:
                playing = False
                continue

            winners[current_friend] = True
            current_friend = (current_friend + current_k * k) % n
            current_k += 1

        losers = []
        for i, player in enumerate(winners):
            if not player:
                losers.append(i)

        return losers

File: leetcode/2682/find_losers_circular_game.cpp
// 6ms | 11mb
// 59.26% | 86.11%
class Solution {
public:
    vector<int> circularGameLosers(int n, int k) {
        vector<bool> winners;
        for (int i = 0; i < n; i++) {
            winners.push_back(false);
        }

        int current_friend = 0;
        int current_k = 1;
        bool playing = true;

        while (playing) {
            if (winners[current_friend]) {
                playing = false;
            } else {
                winners[current_friend] = true;
                current_friend = (current_friend + current_k * k) % n;
                current_k++;
            }
        }

        vector<int> losers;
        for (int i = 0; i < n; i++) {
            if (!winners[i]) {
                losers.push_back(i+1);
            }
        }

        return losers;
    }
};
